maltoolbox/attackgraph/attackgraph.py

```diff
@@ -126,22 +127,21 @@ def _process_step_expression(
             return (new_target_assets, None)
 
         case 'transitive':
-            # The transitive expression is very similar to the field
-            # expression, but it proceeds recursively until no target is
-            # found and it and it sets the new targets to the entire list
-            # of assets identified during the entire transitive recursion.
-            new_target_assets = []
+            # The transitive expression is very similar to the field expression,
+            # but it proceeds recursively until no target is found and it sets
+            # the new targets to the entire list of assets identified during the
+            # entire transitive recursion.
+            new_target_assets = target_assets
```

Comment:
Is this line correct? Do the target_assets by any chance get returned back from the transitive expression which would have them end up being included twice?

Reply:
Yes, that is a potential problem, but it can only happen if an asset reference back to itself directly or indirectly. What we want is to turn this into a set because otherwise we would have to check every time for the `transitive` and field `type` step expressions.

The catch there is that we need to make assets hashable, whcih we probably can since we can just make hashes of their ids, since those are unique.

Reply:
Yes, using the id for the hash makes sense. If an asset is referenced back, won't this cause infinite loops?

Reply:
Yes, we would need to have a check to see if the `new_target_assets` and `target_assets` differ before allowing further recursion. That's really only a problem for the `field` anyway, since `transitive` already has that issue, it has to know when to stop, and that is when it gets no new assets, so we basically need the check from `transitive` also apply for `field` step expressions.

Reply:
Is it good enough as it is here? Maybe only the new lang graph goes to such lengths?

Reply:
But we're not merging this pull request anyway, right? This is just here to help out with updates in the new language graph integration. Or did I misunderstand its purpose?